Instantiate and register split brain messages listener once for the lifetime of a node

vbekiaris opened the pull request:
This PR promotes the existing anonymous listener in `searchForOtherClusters` to top-level class and instantiates & registers the listener just once, during `MulticastJoiner` construction.

Reasoning:
* do not miss any split-brain join messages sent while listener is not registered
* do not produce litter on each execution of split brain handler

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/5937/

vbekiaris commented:
PR builder reports compilation errors in aggregations test:
```
02:41:54 [ERROR] /home/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/test/java/com/hazelcast/aggregation/SumAggregationTest.java:[159,49] type parameters of <T>T cannot be determined; no unique maximal instance exists for type variable T with upper bounds long,java.lang.Number
02:41:54 [ERROR] /home/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/test/java/com/hazelcast/aggregation/SumAggregationTest.java:[187,49] type parameters of <T>T cannot be determined; no unique maximal instance exists for type variable T with upper bounds long,java.lang.Number
02:41:54 [ERROR] /home/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/test/java/com/hazelcast/aggregation/SumAggregationTest.java:[219,51] type parameters of <T>T cannot be determined; no unique maximal instance exists for type variable T with upper bounds double,java.lang.Number
02:41:54 [ERROR] /home/jenkins/jenkins/workspace/new-lab-fast-pr/hazelcast/src/test/java/com/hazelcast/aggregation/SumAggregationTest.java:[251,51] type parameters of <T>T cannot be determined; no unique maximal instance exists for type variable T with upper bounds double,java.lang.Number
```

tombujok commented:
run-lab-run

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/5939/

tombujok commented:
@Donnerbart @vbekiaris There have been some changes in the aggregation tests with a direct commit and no review? 
See the history: https://github.com/hazelcast/hazelcast/commits/master/hazelcast/src/test/java/com/hazelcast/aggregation/SumAggregationTest.java

Donnerbart commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

vbekiaris commented:
Thanks @tombujok @Donnerbart .
Updated to use a `BlockingDeque` as a stack so the split brain join message processed will always be the latest received: the listener adds each message received to the deque's head, split brain job polls from the head; each time split brain job execution ends, the deque is cleared.

devOpsHazelcast commented:
Test PASSed.

tombujok commented on hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MulticastJoiner.java:
```diff
@@ -148,7 +141,7 @@ public void onMessage(Object msg) {
             }
         } finally {
-            node.multicastService.removeMulticastListener(listener);
+            splitBrainJoinMessages.clear();
```
I'm just thinking what if there's more brains than two, for example five? 
With this logic their join requests will be discarded. 
That is the reason I was thinking that we maybe need some message timestamp and expiration duration?

mdogan replied:
As I understand, having `splitBrainJoinMessages.clear()` at the end of each cycle, join messages will be alive at most one cycle of split discovery.

mdogan replied:
I cannot think any problem with processing each message. There are already checks/validations for merge process. This is only discovery part.

vbekiaris replied:
I think expiration is not required. If an old split brain join message is processed, then either the clusters should still merge (win), or if some other changes have occurred in the meanwhile (eg already merged so sending address is not still the master of other cluster, cluster version changed to incompatible version on one cluster) then there will be no merge.

mdogan commented on hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/SplitBrainMulticastListener.java:
```diff
@@ -0,0 +1,46 @@
+    public void onMessage(Object msg) {
+        if (msg != null && msg instanceof SplitBrainJoinMessage) {
+            SplitBrainJoinMessage joinRequest = (SplitBrainJoinMessage) msg;
+            if (node.getThisAddress() != null && !node.getThisAddress().equals(joinRequest.getAddress())) {
```
`node.getThisAddress() != null` is redundant, `thisAddress` cannot be null.

mdogan replied:
`msg != null` is also redundant. `msg instanceof SplitBrainJoinMessage` already implies that.

vbekiaris replied:
Thanks, fixed

vbekiaris commented:
Updated to process all split brain join messages in the stack on each cyle of the split brain handler job.

devOpsHazelcast commented:
Test `FAIL`ed.
https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/5962/

vbekiaris commented:
Unrelated test failure in `SecondsBasedEntryTaskSchedulerTest.test_doNotRescheduleEntryWithinSameSecond_postpone` ( #8728 ).

vbekiaris commented:
run-lab-run

devOpsHazelcast commented:
Test PASSed.

vbekiaris commented:
Thanks @mdogan & @tombujok for your feedback
